internal/contentmanager/cache/cache.go

```diff
@@ -0,0 +1,256 @@
+	}
+	return c.startNewSources(ctx, gvkSet, watcher)
+}
+
+func (c *cache) Close() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	errs := []error{}
+	for _, source := range c.sources {
+		if err := source.Close(); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errors.Join(errs...)
+}
+
+func (c *cache) startNewSources(ctx context.Context, gvks sets.Set[schema.GroupVersionKind], watcher Watcher) error {
+	cacheGvks := sets.New[schema.GroupVersionKind]()
+	for gvk := range c.getSources() {
+		cacheGvks.Insert(gvk)
+	}
+	gvksToCreate := gvks.Difference(cacheGvks)
+
+	errs := make(chan error)
+	wg := sync.WaitGroup{}
+	for _, gvk := range gvksToCreate.UnsortedList() {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			err := c.startNewSource(ctx, gvk, watcher)
+			if err != nil {
+				errs <- err
+			}
+		}()
+	}
+
+	// watch for errors coming through the error channel and append them to the
+	// list of source errors until the error channel is closed
+	errWriteWg := sync.WaitGroup{}
+	errWriteWg.Add(1)
+	sourcesErrors := []error{}
+	go func() {
+		defer errWriteWg.Done()
+		for err := range errs {
+			sourcesErrors = append(sourcesErrors, err)
+		}
+	}()
+
+	// Wait for all sources to finish starting up or erroring out
+	wg.Wait()
+	close(errs)
+
+	// Wait for our error writing goroutine to finish writing any
+	// syncing errors.
+	errWriteWg.Wait()
+
+	return errors.Join(sourcesErrors...)
+}
+
+func (c *cache) startNewSource(ctx context.Context, gvk schema.GroupVersionKind, watcher Watcher) error {
+	s, err := c.sourcerer.Source(gvk, c.owner, func(ctx context.Context) {
+		err := c.removeSource(gvk)
+		if err != nil {
+			logr := log.FromContext(ctx)
+			logr.Error(err, "managed content cache postSyncError removing source failed", "gvk", gvk)
+		}
+	})
+	if err != nil {
+		return fmt.Errorf("getting source: %w", err)
+	}
+
+	err = watcher.Watch(s)
+	if err != nil {
+		return fmt.Errorf("establishing watch for GVK %q: %w", gvk, err)
+	}
+
+	syncCtx, syncCancel := context.WithTimeout(ctx, c.syncTimeout)
+	defer syncCancel()
+	err = s.WaitForSync(syncCtx)
+	if err != nil {
+		return err
+	}
+
+	err = c.addSource(gvk, s)
+	if err != nil {
+		// If we made it here then there is a logic error in
+		// calculating the diffs between what is currently being
+		// watched by the cache
+		panic(err)
+	}
+
+	return nil
+}
+
+func (c *cache) addSource(gvk schema.GroupVersionKind, source CloserSyncingSource) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if _, ok := c.sources[gvk]; !ok {
+		c.sources[gvk] = source
+		return nil
+	}
+	return errors.New("source already exists")
+}
+
+func (c *cache) removeStaleSources(gvks sets.Set[schema.GroupVersionKind]) error {
+	cacheGvks := sets.New[schema.GroupVersionKind]()
+	for gvk := range c.getSources() {
+		cacheGvks.Insert(gvk)
+	}
+
+	removeErrs := []error{}
+	gvksToRemove := cacheGvks.Difference(gvks)
+	for _, gvk := range gvksToRemove.UnsortedList() {
+		err := c.removeSource(gvk)
+		if err != nil {
```

Comment:
Total nit, but this is philosophical: we _could_ collect nil errors into the list because `errors.Join` will ignore them.

Reply:
Going to leave as is because collection of nil errors may cause confusion for readers unfamiliar with the `errors.Join` function